initial pass at documenting "writing tests" (#115)
BrightspaceUI/testing: README.md

[![NPM version](https://img.shields.io/npm/v/@brightspace-ui/testing.svg)](https://www.npmjs.org/package/@brightspace-ui/testing)

> Testing utilities which are specifically designed and configured for Brightspace UI components and applications.

Author: I don't want to take away that this is building on other stuff, but I moved this information down to where it's most relevant.

npm install @brightspace-ui/testing
```

## Writing Tests

Tests leverage the familiar [Mocha](https://mochajs.org/) and [Chai](https://www.chaijs.com/) test frameworks. Many of the available utilities are wrappers around or extensions of the excellent [@open-wc testing helpers](https://open-wc.org/docs/testing/helpers/).

### describe, it, assert & expect

Tests can be grouped into suites using `describe` and are defined inside `it` blocks.

> Results can be verified using either the BDD-style `expect` or TDD-style `assert` syntax (although try not to mix & match).

Reviewer: > (although try not to mix & match)

Lol and then you do just that. Maybe broken into two different code blocks?
```javascript
import { expect } from '@brightspace-ui/testing';

describe('group of tests', () => {
  it('should multiply numbers', () => {
    expect(2 * 4).to.equal(8);
  });
});
```
```javascript
import { assert} from '@brightspace-ui/testing';

describe('group of tests', () => {
  it('should multiply numbers', () => {
    assert.equal(2 * 4, 8);
  });
});
```

Maybe that's overkill

});
```

Each call to `fixture()` will restore the browser to its default state, making subsequent calls isolated from each other. There's therefore no need to manually restore the viewport, language, mouse, or keyboard in between tests.

#### Configuring the Viewport Size

The viewport defaults to `800px` wide by `800px` tall. To use different viewport sizes, pass a `viewport` option to `fixture()`.

```javascript
it('should work on small viewports', async() => {
  const elem = await fixture(html`<my-elem></my-elem>`, {
    viewport: { height: 300, width: 200 }
  });
  // do assertions
});
```

#### Configuring the Language or Text Direction

If the component under test has special multi-lingual or bidirectional text behavior, both `language` and `rtl` (right-to-left) options are available.

```javascript
it('should work in French', async() => {
  const elem = await fixture(html`<my-elem></my-elem>`, {
    lang: 'fr'
  });
  // do assertions
});

it('should work in RTL', async() => {
  const elem = await fixture(html`<my-elem></my-elem>`, {
    rtl: true
  });
  // do assertions
});
```

If `lang` is set to Arabic (`ar`), the right-to-left option will automatically be enabled.

> > **Note:** it's not recommended to use `language` configuration with visual-diff to solely test the correctness of translations. The [messageformat-validator](https://github.com/bearfriend/messageformat-validator) is a more efficient way to test translations.

Reviewer: Lol we should probably get that moved into `BrightspaceUI` at some point

Author: Yeah, or even into this library! `d2l-test-mfv`

```javascript
import { clickElem, fixture, hoverElem } from '@brightspace-ui/testing';

it('should hover over element', async() => {
  const elem = await fixture(...);
  await hoverElem(elem);
  // do assertions
});

it('should click on element', async() => {
  const elem = await fixture(...);
  await clickElem(elem);
  // do assertions
});
```

Alternatively, to hover over or click at viewport coordinates with the mouse, use `hoverAt(x, y)` or `clickAt(x, y)`:

```javascript
import { clickAt, hoverAt } from '@brightspace-ui/testing';

it('should hover at coordinate', async() => {
  await hoverAt(100, 200);
  // do assertions
});

it('should click at coordinate', async() => {
  await clickAt(100, 200);
  // do assertions
});
```

### Using the Keyboard

To place focus on an element using the keyboard, use `focusElem(elem)`. Doing so will trigger its `:focus-visible` CSS pseudo-class.

```javascript
import { fixture, focusElem } from '@brightspace-ui/testing';

it('should focus on element', async() => {
  const elem = await fixture(html`<button>elem</button>`);
  await focusElem(elem);
  // do assertions
});
```

> To send particular keystrokes to the browser window or a specific element, use `sendKeys(action, keys)` or `sendKeysElem(elem, action, keys)`.

Reviewer: I think it'd be worth calling out somewhere in here that `sendKeysElem` will focus the element first, not just fire a keyboard event from it

it('should wait for condition', async() => {
  const elem = await fixture(...);
  await waitUntil(() => elem.foo === 'bar');
});
```

By default, `waitUntil` will poll every `50ms` and time out after `1000ms`. Those options can be configured:

```javascript
await waitUntil(() => elem.condition, {
  interval: 10,
  timeout: 2000
});
```

> **Note:** because `waitUntil` constantly polls, it can slow down test execution and should be avoided if possible.

### Defining a Custom Element for Testing

If a test requires a one-off custom element, define it using `defineCE` and pass the returned tag name in the call to `fixture()`.

```javascript
import { defineCE, fixture, html } from '@brightspace-ui/testing';

const tag = defineCE(
  class extends LitElement {
    static properties = {
      foo: { type: String }
    };
    render() {
      return html`hello`;
    }
  }
);

it('should use custom element', async() => {
  const foo = 'bar';
  const elem = await fixture(html`<${tag} foo="${foo}"></${tag}>`);
});
```

> **Important:** `defineCE` is not performant and shouldn't be used outside of test files.

### Visual-Diff Testing

> Also known as "visual regression" or "perceptual diff", visual-diff testing involves taking snapshot images of the browser and comparing them against a known golden (or "baseline") image. The images are compared pixel-by-pixel and differences beyond a threshold will fail the test. Our visual diff testing leverages the [pixelmatch](https://github.com/mapbox/pixelmatch) library to perfom its comparison.

Reviewer: Maybe worth calling it `vdiff` in at least a few places here, or everywhere?  That's what we've been using everywhere else, in the command, action, test names, etc.

Use the asynchronous `.to.be.golden()` Chai assertion to take a visual-diff snapshot and compare it against its golden.

```javascript
import { fixture, html } from '@brightspace-ui/testing';

describe('my-elem', () => {
  describe('situation1', () => {
    it('state1', async() => {
      const elem = await fixture(html`<my-elem></my-elem>`);
      await expect(elem).to.be.golden();
    });
  });
});
```

> The filename and location of the resulting image will be based on the suite names and test name. The top-most suite will become a directory, and any remaining suites will be combined with the test name into the file name. In this example, the snapshot would be stored in: `./my-elem/situation1-state1.png`

Reviewer: Are we going to have a separate section with more info on `vdiff`?  This path isn't really accurate, for local or for ci, and I figured we'd talk about what file names are expected, link to the action, etc.

#### Configuring the Snapshot Area

By default, the snapshot area will be a rectangle around the source element plus a `10px` buffer margin on each side. To use a different margin, pass it as an option:

```javascript
await expect(elem).to.be.golden({ margin: 20 });
```

## Running Tests

TODO: note about leveraging @web/test-runner

## Debugging Tests

TODO

## Developing & Contributing

After cloning the repo, run `npm install` to install its dependencies.

### Testing

To run the full suite of tests:

```shell
npm test
```

Alternately, tests can be selectively run:

```shell
# binary unit tests
npm run test:bin

# browser unit tests
npm run test:browser

# server unit tests
npm run test:server

# vdiff tests
npm run test:vdiff
```

### Versioning & Releasing

> This repo is configured to use the `semantic-release`. Commits prefixed with `fix:` and `feat:` will trigger patch and minor releases when merged to `main`.

Reviewer: ```suggestion
This repo is configured to use the `semantic-release` action. Commits prefixed with `fix:` and `feat:` will trigger patch and minor releases when merged to `main`.
```
```suggestion
This repo is configured to use `semantic-release`. Commits prefixed with `fix:` and `feat:` will trigger patch and minor releases when merged to `main`.
```

* `down`: holds down a single key
* `up`: releases a single key

For a list of all available key values, refer to [key values for keyboard events](https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_key_values).

```javascript
import { fixture, sendKeysElem } from '@brightspace-ui/testing';

it('should type and press Enter', async() => {
  const elem = await fixture(html`<input type="text">`);
  await sendKeysElem(elem, 'type', 'Hello');
  await sendKeysElem(elem, 'press', 'Enter');
});
```

As demonstrated above, subsequent calls can be made to create key combinations.

### Waiting for Things

There are various scenarios where a test may need to wait before it can proceed.

#### Waiting for an Event

After interacting with components, to wait for a particular event to be dispatched use `oneEvent(elem, eventName)`.

```javascript
import { clickElem, fixture, oneEvent } from '@brightspace-ui/testing';

it('should wait for an event', async() => {
  const elem = await fixture(html`<my-elem></my-elem>`);
  clickElem(elem);
  await oneEvent(elem, 'some-event');
});
```

> Note that the call to `clickElem` is not `await`-ed, since by the time it resolves the event will likely have already been dispatched.

Reviewer 2: I feel like it's clearer to me to state it the other way around.
```suggestion
Note that the call to `clickElem` is not `await`-ed, since by the time it resolves the event will have already been dispatched.
```
```suggestion
Note that the call to `clickElem` is not `await`-ed, so that the listener is registered before the event is dispatched.
```
Fine either way though. Whatever people prefer.

An even safer approach would be to wrap that call in a `setTimeout`:

```javascript
setTimeout(() => clickElem(elem));
await oneEvent(elem, 'some-event');
```

> #### Waiting for a Lit Element to Update

Author: This could easily be left out, but it could be useful to call it out in case folks use `aTimeout` or `nextFrame` instead of realizing they can do this? The [open-wc stuff](https://open-wc.org/docs/testing/helpers/#elementupdated) has a special `await elementUpdated(elem)` that essentially does this and falls back to `nextFrame` if it's not a Lit element.

> When using Lit-based components, it's common to create a fixture, modify some properties and then want to wait for those changes to be rendered.

Reviewer 2: Just seems clearer to me
```suggestion
When using Lit-based components, it's common to create a fixture and then modify some of its properties, which usually requires waiting for those changes to be rendered.
```

```javascript
import { fixture, waitUntil } from '@brightspace-ui/testing';

it('should wait for condition', async() => {
	const elem = await fixture(...);
	await waitUntil(() => elem.foo === 'bar');
})
```

By default, `waitUntil` will poll every `50ms` and time out after `1000ms`. Those options can be configured:

```javascript
await waitUntil(() => elem.condition, {
  interval: 10,
  timeout: 2000
});
```

> **Note:** because `waitUntil` constantly polls, it can slow down test execution and should be avoided if possible.

### Defining a Custom Element for Testing

If a test requires a one-off custom element, define it using `defineCE` and pass the returned tag name in the call to `fixture()`.

```javascript
import { defineCE, fixture, html } from '@brightspace-ui/testing';

const tag = defineCE(
  class extends LitElement {
    static properties = {
      foo: { type: String }
    }
    render() {
      return html`hello`;
    }
  }
);

it('should use custom element', async() => {
  const foo = 'bar';
  const elem = await fixture(html`<${tag} foo="${foo}"></${tag}>`);
});
```

> ## Versioning & Releasing

Author: The diff got pretty confused, but I've trimmed down our "semantic-release" blurb significantly since it repeats so much information that's already in the action README. If folks like this, I'd like to make it the template for all repos.

> > **Important:** `defineCE` is not performant and shouldn't be used outside of tests.

Reviewer 2: ```suggestion
> **Important:** `defineCE` is not performant and shouldn't be used outside of test files.
```

Reviewer 2: Oh, or does this mean used the result of `defineCE` in tests?

Author: From the open-wc docs:
> This uses a workaround that's not performant for re-renders, which is fine for testing, but do NOT use this in production code.

Reviewer 2: Ok, then I still with the suggestion above for clarity since `defineCE` is usually used outside of the literal `it` test, but always only in test files.

> Unlike other tests, visual-diff tests must be in a file with the `*.vdiff.js` extension.

Reviewer 2: ```suggestion
Unlike other tests, visual-diff tests must be in a file that matches the pattern `*.vdiff.js`.
```

Reviewer 2: Actually, this isn't quite accurate either since they can defines their own pattern. It just uses `vdiff` as the type _in_ that pattern.

Author: Right. So should I just leave this part totally off and it can be discussed down in the "running tests" part?

Reviewer 2: Yeah maybe that's best

> import { fixture, html } from '@brightspace-ui/testing';

Author: _Possibly_, yeah. Although really visual-diff is just getting in line with our other "unit" tests which always had to do those imports. I'll think about this a bit more.
